src/types/list.rs

```diff
@@ -76,23 +80,130 @@ impl PyList {
     /// All standard library structures implement this trait correctly, if they do, so calling this
     /// function with (for example) [`Vec`]`<T>` or `&[T]` will always succeed.
     #[track_caller]
-    pub fn new<T, U>(py: Python<'_>, elements: impl IntoIterator<Item = T, IntoIter = U>) -> &PyList
+    pub fn new<'py, T, U>(
+        py: Python<'py>,
+        elements: impl IntoIterator<Item = T, IntoIter = U>,
+    ) -> Py2<'py, PyList>
     where
         T: ToPyObject,
         U: ExactSizeIterator<Item = T>,
     {
         let mut iter = elements.into_iter().map(|e| e.to_object(py));
-        let list = new_from_iter(py, &mut iter);
-        list.into_ref(py)
+        new_from_iter(py, &mut iter)
     }
 
     /// Constructs a new empty list.
-    pub fn empty(py: Python<'_>) -> &PyList {
-        unsafe { py.from_owned_ptr::<PyList>(ffi::PyList_New(0)) }
+    pub fn empty(py: Python<'_>) -> Py2<'_, PyList> {
+        unsafe { Py2::from_owned_ptr(py, ffi::PyList_New(0)).downcast_into_unchecked() }
     }
+}
 
+/// Implementation of functionality for [`PyList`].
+///
+/// These methods are defined for the `Py2<'py, PyList>` smart pointer, so to use method call
+/// syntax these methods are separated into a trait, because stable Rust does not yet support
+/// `arbitrary_self_types`.
+pub trait PyListMethods<'py> {
```

Comment:
The trait appears to be a nice trick to work around the documentation issues of inherent implementations on `Py2<'py, PyList>`. Maybe discoverability could be improved using the [`#[doc(alias = "PyList")]` feature](https://doc.rust-lang.org/rustdoc/write-documentation/the-doc-attribute.html#alias) of Rustdoc?

Two things I wonder though: Is there a cost in compile time for this additional indirection? Also, should we keep the traits open or seal them to be able to make additional assumptions in their definition and/or implementation?

Reply:
Personally I think sealing them makes sense, seems like only trouble could be had by users implementing them.

Reply:
Added a commit to push doc aliases, that seems like a good idea and easy to implement.

Regarding the further questions:

- For the additional indirection: I think although it's likely to be a more complex lookup than inherent methods, the is hopefully negligible. We use trait methods like `.to_string()` and `.clone()`  all over Rust code so I think if it was not efficient to resolve these names there would be wide recognition of a problem. Or was there a different concern you had wrt compile times?
- Sealing these traits: yes I think it cannot hurt and can always be reversed later more easily than sealing later.